README.md
# mods

Functions that implement equations from "Mechanics of Materials" by R.C. Hibbeler (9th edition).

![Major principal stress](images/readme_image_1.png)

## Installation

The following does not yet work as this package is not yet on PyPI.

```bash
$ pip install mods
```

For now, perform the following steps:

1. Clone the repo
2. `cd` into the cloned repository and do:

```bash
$ poetry install
```

You need to ensure that you have `poetry` installed and accessible from your command line. Ideally, you should be
working in a `conda` environment.

## Documentation

You can find the documentation [here](https://mechanics-of-deformable-solids.readthedocs.io/).

## Usage

You can plot Mohr's circle by providing `sigma_x`, `sigma_y`, `tau_xy` and a plane inclination angle `theta` (in degrees):

```python
from mods.stress_transformation import mohrs_circle_plane_angle_plot

sigma_x = 1
sigma_y = 3
tau_xy = 2
theta = 30

mohrs_circle_plane_angle_plot(sigma_x, sigma_x, tau_xy, theta)
```

![](images/mohrs_circle_plane_angle_plot.png)

See the [documentation](https://mechanics-of-deformable-solids.readthedocs.io/) for more details.

## About the name
"mods" stands for "mechanics of deformable solids", which is the course title for ME 219 and ME 220
at the University of Waterloo's department of mechanical and mechatronics engineering. "Mechanics of Materials"
by Hibbeler is the textbook used in these two courses.

## Contributing

Interested in contributing? Check out the contributing guidelines. Please note that this project is released with a Code of Conduct. By contributing to this project, you agree to abide by its terms.

## License

`mods` was created by John You. It is licensed under the terms of the MIT license.

## Credits

`mods` was created with [`cookiecutter`](https://cookiecutter.readthedocs.io/en/latest/) and the `py-pkgs-cookiecutter` [template](https://github.com/py-pkgs/py-pkgs-cookiecutter).


CHANGELOG.md
# Changelog

<!--next-version-placeholder-->

## v0.1.0 (28/04/2023)

- First release of `mods`!

tests/test_mods.py
from mods import stress_transformation


src/mods/mods.py
import math

def normal_stress_transform(sigma_x, sigma_y, tau_xy, theta):
    """
    Calculate the normal stress on an inclined plane.

    Parameters:
    sigma_x (float): Normal stress in the x direction.
    sigma_y (float): Normal stress in the y direction.
    tau_xy (float): Shear stress in the x-y plane.
    theta (float): Angle of the inclined plane in degrees.

    Returns:
    float: Normal stress on the inclined plane.
    """
    theta_rad = math.radians(theta)  # converting degrees to radians
    sigma_n = (sigma_x + sigma_y)/2 + (sigma_x - sigma_y)/2*math.cos(2*theta_rad) + tau_xy*math.sin(2*theta_rad)
    return sigma_n

def shear_stress_transform(sigma_x, sigma_y, tau_xy, theta):
    """
    Calculate the shear stress on an inclined plane.

    Parameters:
    sigma_x (float): Normal stress in the x direction.
    sigma_y (float): Normal stress in the y direction.
    tau_xy (float): Shear stress in the x-y plane.
    theta (float): Angle of the inclined plane in degrees.

    Returns:
    float: Shear stress on the inclined plane.
    """
    theta_rad = math.radians(theta)  # converting degrees to radians
    tau_n = -(sigma_x - sigma_y)/2*math.sin(2*theta_rad) + tau_xy*math.cos(2*theta_rad)
    return tau_n

def principal_stress(sigma_x, sigma_y, tau_xy, stress_type):
    """
    Calculate the principal stress.

    Parameters:
    sigma_x (float): Normal stress in the x direction.
    sigma_y (float): Normal stress in the y direction.
    tau_xy (float): Shear stress in the x-y plane.
    stress_type (str): 'max' for major principal stress, 'min' for minor principal stress.

    Returns:
    float: Principal stress, either major or minor based on stress_type.

    Raises:
    ValueError: If stress_type is not 'max' or 'min'.
    """
    if stress_type == "max":
        sigma = (sigma_x + sigma_y)/2 + math.sqrt(((sigma_x - sigma_y)/2)**2 + tau_xy**2)
    elif stress_type == "min":
        sigma = (sigma_x + sigma_y)/2 - math.sqrt(((sigma_x - sigma_y)/2)**2 + tau_xy**2)
    else:
        raise ValueError("Invalid stress_type. Choose either 'max' or 'min'.")
    return sigma

def principal_stress_angle(sigma_x, sigma_y, tau_xy):
    """
    Calculate the principal stress angle.

    Parameters:
    sigma_x (float): Normal stress in the x direction.
    sigma_y (float): Normal stress in the y direction.
    tau_xy (float): Shear stress in the x-y plane.

    Returns:
    float: Principal stress angle in degrees.
    """
    numerator = 2 * tau_xy
    denominator = sigma_x - sigma_y
    theta_p_rad = math.atan2(numerator, denominator) / 2  # atan2 takes care of the angle quadrant
    theta_p_deg = math.degrees(theta_p_rad)  # converting radians to degrees
    return theta_p_deg


src/mods/failure.py
import numpy as np
import matplotlib.pyplot as plt
from matplotlib.patches import Ellipse
from mods.stress_transformation import principal_stress

def plot_von_mises_failure_envelope(sy, sigma_x=None, sigma_y=None, tau_xy=None):
    """
    Plots the Von Mises failure envelope for plane stress.

    Parameters:
        sy (float): Yield strength of the material
        sigma_x (float): Stress value in the x-direction
        sigma_y (float): Stress value in the y-direction
        tau_xy (float): Shear stress value in the xy-plane

    Returns:
        None

    Raises:
        ValueError: If all three stress values are not provided
    """
    # Sy is the yield strength of the material
    ellipse = Ellipse((0, 0), width=sy*np.sqrt(2)*2, height=sy*np.sqrt(2/3)*2, angle=45, alpha=0.4)

    fig, ax = plt.subplots(figsize=(8, 8))
    ax.add_patch(ellipse)
    ax.set_xlim(-sy*np.sqrt(2), sy*np.sqrt(2))
    ax.set_ylim(-sy*np.sqrt(2), sy*np.sqrt(2))
    ax.set_aspect('equal', adjustable='box')

    # Draw bold black lines at x=0 and y=0
    ax.axhline(0, color='black', linewidth=1.5)
    ax.axvline(0, color='black', linewidth=1.5)

    # If stress values are given, calculate and plot the principal stresses
    if sigma_x is not None and sigma_y is not None and tau_xy is not None:
        sigma_1, sigma_2 = principal_stress(sigma_x, sigma_y, tau_xy)
        ax.plot(sigma_1, sigma_2, 'ro', markersize=8)
    elif sigma_x is not None or sigma_y is not None or tau_xy is not None:
        raise ValueError("Please provide all three stress values: sigma_x, sigma_y, and tau_xy")
    
    plt.title("Von Mises Failure Envelope for Plane Stress")
    plt.xlabel("Major Principal Stress (σ1)")
    plt.ylabel("Minor Principal Stress (σ2)")
    plt.grid(True)
    plt.show()


src/mods/buckling.py
import math

def max_stress(P, A, e, c, r, L, E):
    """
    Calculates the maximum stress experienced by a column using the secant formula.

    Parameters:
        P (float): Applied load
        A (float): Cross-sectional area of the column
        e (float): Eccentricity of the load
        c (float): Distance from the centroid of the cross-section to the extreme fiber
        r (float): Radius of gyration of the cross-section
        L (float): Length of the column
        E (float): Modulus of elasticity of the material

    Returns:
        float: Maximum stress experienced by the column
    """
    sec_term = math.sec((L/(2*r)) * math.sqrt(P/(E*A)))
    max_stress = (P/A) * (1 + (e*c)/(r**2) * sec_term)
    return max_stress

def critical_load(E, I, L, support):
    """
    Calculates the critical load on an ideal column using Euler's formula.
    This is the maximum axial load that a member can carry before buckling.

    Parameters:
        E (float): Modulus of elasticity of the material
        I (float): Moment of inertia of the cross-section of the column
        L (float): Length of the column
        support (str): Type of support, can be 'pin', 'fixed', 'pin-fixed', or 'fixed-free'

    Returns:
        float: Critical load on the column
    """
    K_dict = {'pin': 1, 'fixed': 0.5, 'pin-fixed': 0.7, 'fixed-free': 2}
    K = K_dict[support]

    P_cr = (math.pi**2 * E * I)/(K * L)**2

    return P_cr

def critical_stress(E_t, L, r, support):
    """
    Calculates the critical stress on an ideal column using Engesser's equation with tangent modulus.

    Parameters:
        E_t (float): Tangent modulus of elasticity of the material
        L (float): Length of the column
        r (float): Radius of gyration of the cross-section of the column
        support (str): Type of support, can be 'pin', 'fixed', 'pin-fixed', or 'fixed-free'

    Returns:
        float: Critical stress on the column
    """
    K_dict = {'pin': 1, 'fixed': 0.5, 'pin-fixed': 0.7, 'fixed-free': 2}
    K = K_dict[support]

    sigma_cr = (math.pi**2 * E_t)/((K * L/r)**2)

    return sigma_cr


src/mods/datasets.py
from importlib import resources
import pandas as pd

def get_mechanical_properties_si():
    """Get path to 'avg_mech_prop_si.csv'

    Returns
    -------
    pandas.DataFrame

    References
    ----------
    [1] R.C. Hibbeler, Mechanics of Materials, 9th ed., Pearson, 2014.
    """
    with resources.path("mods.data", "avg_mech_prop_si.csv") as f:
        data_file_path = f
    df = pd.read_csv(data_file_path)
    return df

def get_mechanical_properties_imperial():
    """Get path to 'avg_mech_prop_imperial.csv'

    Returns
    -------
    pandas.DataFrame

    References
    ----------
    [1] R.C. Hibbeler, Mechanics of Materials, 9th ed., Pearson, 2014.
    """
    with resources.path("mods.data", "avg_mech_prop_imperial.csv") as f:
        data_file_path = f
    df = pd.read_csv(data_file_path)
    return df

def get_w_shapes_imperial():
    """Get path to 'w_shapes_fps_imperial.csv'

    Returns
    -------
    pandas.DataFrame

    References
    ----------
    [1] R.C. Hibbeler, Mechanics of Materials, 9th ed., Pearson, 2014.
    """
    with resources.path("mods.data", "w_shapes_fps_imperial.csv") as f:
        data_file_path = f
    df = pd.read_csv(data_file_path)
    return df

def get_w_shapes_si():
    """Get path to 'w_shapes_fps_si.csv'

    Returns
    -------
    pandas.DataFrame

    References
    ----------
    [1] R.C. Hibbeler, Mechanics of Materials, 9th ed., Pearson, 2014.
    """
    with resources.path("mods.data", "w_shapes_fps_si.csv") as f:
        data_file_path = f
    df = pd.read_csv(data_file_path)
    return df


src/mods/strain_transformation.py
import math
import matplotlib.pyplot as plt

def average_strain(epsilon_x, epsilon_y):
    """
    Calculates the average strain and returns the value.

    Parameters:
        epsilon_x (float): Strain value in the x-direction
        epsilon_y (float): Strain value in the y-direction

    Returns:
        float: Average strain value
    """
    epsilon_avg = (epsilon_x + epsilon_y)/2

    return epsilon_avg

def normal_strain_transform(epsilon_x, epsilon_y, gamma_xy, theta):
    """
    Applies a strain transformation to the given strains and returns the transformed strain value.

    Parameters:
            epsilon_x (float): Strain value in the x-direction
            epsilon_y (float): Strain value in the y-direction
            gamma_xy (float): Shear strain value in the xy-plane
            theta (float): Angle of rotation in degrees

    Returns:
        tuple: Transformed strain value in the x'-direction and y'-direction
    """
    theta_rad = math.radians(theta)
    cos_theta = math.cos(2*theta_rad)
    sin_theta = math.sin(2*theta_rad)

    epsilon_x_prime = (epsilon_x + epsilon_y)/2 + (epsilon_x - epsilon_y)/2 * cos_theta + gamma_xy/2 * sin_theta
    epsilon_y_prime = (epsilon_x + epsilon_y)/2 - (epsilon_x - epsilon_y)/2 * cos_theta - gamma_xy/2 * sin_theta

    return epsilon_x_prime, epsilon_y_prime

def shear_strain_transform(epsilon_x, epsilon_y, gamma_xy, theta):
    """
    Applies a transformation to the given shear strain and returns the transformed shear strain value.

    Parameters:
        epsilon_x (float): Strain value in the x-direction
        epsilon_y (float): Strain value in the y-direction
        gamma_xy (float): Shear strain value in the xy-plane
        theta (float): Angle of rotation in degrees

    Returns:
        float: Transformed shear strain value in the x'y'-plane
    """
    theta_rad = math.radians(theta)
    cos_theta = math.cos(2*theta_rad)
    sin_theta = math.sin(2*theta_rad)

    gamma_xy_prime = -1*(epsilon_x - epsilon_y)/2 * sin_theta + gamma_xy/2 * cos_theta

    return gamma_xy_prime

def principal_strain(epsilon_x, epsilon_y, gamma_xy):
    """
    Calculates the principal strains and returns the values in a tuple.

    Parameters:
        epsilon_x (float): Strain value in the x-direction
        epsilon_y (float): Strain value in the y-direction
        gamma_xy (float): Shear strain value in the xy-plane

    Returns:
        tuple: Principal strain values in the x'-direction and y'-direction
    """
    epsilon_prime = (epsilon_x + epsilon_y)/2
    epsilon_double_prime = (epsilon_x - epsilon_y)/2
    gamma_prime = gamma_xy/2

    epsilon_1 = epsilon_prime + math.sqrt(epsilon_double_prime**2 + gamma_prime**2)
    epsilon_2 = epsilon_prime - math.sqrt(epsilon_double_prime**2 + gamma_prime**2)

    return epsilon_1, epsilon_2

def principal_strain_angle(epsilon_x, epsilon_y, gamma_xy):
    """
    Calculates the principal directions and returns the values in a tuple.

    Parameters:
        epsilon_x (float): Strain value in the x-direction
        epsilon_y (float): Strain value in the y-direction
        gamma_xy (float): Shear strain value in the xy-plane

    Returns:
        tuple: Principal direction values in degrees
    """
    epsilon_double_prime = (epsilon_x - epsilon_y)/2
    gamma_prime = gamma_xy/2

    theta_1 = math.degrees(math.atan2(gamma_prime, epsilon_double_prime))/2
    theta_2 = theta_1 + 90

    return theta_1, theta_2

def maximum_in_plane_shear_strain(epsilon_x, epsilon_y, gamma_xy):
    """
    Calculates the maximum in-plane shear strain and returns the value.

    Parameters:
        epsilon_x (float): Strain value in the x-direction
        epsilon_y (float): Strain value in the y-direction
        gamma_xy (float): Shear strain value in the xy-plane

    Returns:
        float: Maximum in-plane shear strain value
    """
    epsilon_double_prime = (epsilon_x - epsilon_y)/2
    gamma_prime = gamma_xy/2

    gamma_max = 2*math.sqrt(epsilon_double_prime**2 + gamma_prime**2)

    return gamma_max

def maximum_in_plane_shear_strain_angle(epsilon_x, epsilon_y, gamma_xy):
    """
    Calculates the angle of the maximum in-plane shear strain and returns the value.

    Parameters:
        epsilon_x (float): Strain value in the x-direction
        epsilon_y (float): Strain value in the y-direction
        gamma_xy (float): Shear strain value in the xy-plane

    Returns:
        float: Angle of the maximum in-plane shear strain value in degrees
    """
    epsilon_double_prime = (epsilon_x - epsilon_y)/2
    gamma_prime = gamma_xy/2

    theta_max = math.degrees(math.atan2(-epsilon_double_prime, gamma_prime))/2

    return theta_max, theta_max + 90

def mohrs_circle(epsilon_x, epsilon_y, gamma_xy):
    """
    Calculates the center and radius of the Mohr's circle and returns the values in a tuple.

    Parameters:
        epsilon_x (float): Strain value in the x-direction
        epsilon_y (float): Strain value in the y-direction
        gamma_xy (float): Shear strain value in the xy-plane

    Returns:
        tuple: Center and radius of the Mohr's circle
    """
    epsilon_prime = (epsilon_x + epsilon_y)/2
    epsilon_double_prime = (epsilon_x - epsilon_y)/2
    gamma_prime = gamma_xy/2

    center = (epsilon_prime, 0)
    radius = math.sqrt(epsilon_double_prime**2 + gamma_prime**2)

    return center, radius

def mohrs_circle_plot(epsilon_x, epsilon_y, gamma_xy):
    """
    Plots the Mohr's circle and returns the plot.

    Parameters:
        epsilon_x (float): Strain value in the x-direction
        epsilon_y (float): Strain value in the y-direction
        gamma_xy (float): Shear strain value in the xy-plane

    Returns:
        matplotlib.pyplot.plot: Plot of the Mohr's circle
    """
    center, radius = mohrs_circle(epsilon_x, epsilon_y, gamma_xy)

    fig, ax = plt.subplots(figsize=(8,8))
    ax.set_aspect('equal')
    ax.grid(True, which='both')
    ax.axhline(y=0, color='k')
    ax.axvline(x=0, color='k')
    ax.set_xlabel(r'$\epsilon$')
    ax.set_ylabel(r'$\gamma$')
    ax.set_title("Mohr's Circle")

    circle = plt.Circle(center, radius, color='b', fill=False)
    ax.add_artist(circle)
    return fig

def mohrs_circle_strain(epsilon_x, epsilon_y, gamma_xy, theta):
    """
    Calculates the center and radius of the transformed Mohr's circle and returns the values in a tuple.

    Parameters:
        epsilon_x (float): Strain value in the x-direction
        epsilon_y (float): Strain value in the y-direction
        gamma_xy (float): Shear strain value in the xy-plane
        theta (float): Angle of rotation in degrees

    Returns:
        tuple: Center and radius of the transformed Mohr's circle
    """
    epsilon_x_prime, epsilon_y_prime = normal_strain_transform(epsilon_x, epsilon_y, gamma_xy, theta)
    gamma_xy_prime = shear_strain_transform(epsilon_x, epsilon_y, gamma_xy, theta)

    return mohrs_circle_strain(epsilon_x_prime, epsilon_y_prime, gamma_xy_prime)

def mohrs_circle_strain_plot(epsilon_x, epsilon_y, gamma_xy, theta):
    """
    Plots the Mohr's circle and a point representing the transformed strain state and returns the plot.

    Parameters:
        epsilon_x (float): Strain value in the x-direction
        epsilon_y (float): Strain value in the y-direction
        gamma_xy (float): Shear strain value in the xy-plane
        theta (float): Plane inclination angle in degrees

    Returns:
        matplotlib.pyplot.plot: Plot of the Mohr's circle and transformed strain state
    """
    epsilon_x_prime, gamma_xy_prime = mohrs_circle_strain(epsilon_x, epsilon_y, gamma_xy, theta)
    center, radius = mohrs_circle(epsilon_x, epsilon_y, gamma_xy)
    circle = plt.Circle((center, 0), radius, fill=False)
    fig, ax = plt.subplots()
    ax.set_aspect('equal')
    ax.add_artist(circle)

    # Set the x and y limits
    ax.set_xlim([center - radius - 0.1 * radius, center + radius + 0.1 * radius])  # adding a little extra space
    ax.set_ylim([-radius - 0.1 * radius, radius + 0.1 * radius])  # adding a little extra space

    plt.plot(epsilon_x_prime, gamma_xy_prime, 'ro')
    plt.xlabel("Normal Strain")
    plt.ylabel("Shear Strain")
    plt.title("Mohr's Circle")
    plt.grid()
    plt.show()
    return fig


src/mods/stress_transformation.py
import math
import matplotlib.pyplot as plt

def average_normal_stress(sigma_x, sigma_y):
    """
    Calculate the average normal stress.

    Parameters:
        sigma_x (float): Normal stress in the x direction.
        sigma_y (float): Normal stress in the y direction.

    Returns:
        float: Average normal stress.
    """
    sigma_avg = (sigma_x + sigma_y)/2
    return sigma_avg

def normal_stress_transform(sigma_x, sigma_y, tau_xy, theta):
    """
    Calculate the normal stress on an inclined plane.

    Parameters:
        sigma_x (float): Normal stress in the x direction.
        sigma_y (float): Normal stress in the y direction.
        tau_xy (float): Shear stress in the x-y plane.
        theta (float): Angle of the inclined plane in degrees.

    Returns:
        tuple: Normal stress on the inclined plane in the x' and y' directions.
    """
    theta_rad = math.radians(theta)  # converting degrees to radians
    sigma_x_prime = (sigma_x + sigma_y)/2 + (sigma_x - sigma_y)/2*math.cos(2*theta_rad) + tau_xy*math.sin(2*theta_rad)
    sigma_y_prime = (sigma_x + sigma_y)/2 - (sigma_x - sigma_y)/2*math.cos(2*theta_rad) - tau_xy*math.sin(2*theta_rad)
    return sigma_x_prime, sigma_y_prime

def shear_stress_transform(sigma_x, sigma_y, tau_xy, theta):
    """
    Calculate the shear stress on an inclined plane.

    Parameters:
        sigma_x (float): Normal stress in the x direction.
        sigma_y (float): Normal stress in the y direction.
        tau_xy (float): Shear stress in the x-y plane.
        theta (float): Angle of the inclined plane in degrees.

    Returns:
        float: Shear stress on the inclined plane.
    """
    theta_rad = math.radians(theta)  # converting degrees to radians
    tau_n = -(sigma_x - sigma_y)/2*math.sin(2*theta_rad) + tau_xy*math.cos(2*theta_rad)
    return tau_n

def principal_stress(sigma_x, sigma_y, tau_xy):
    """
    Calculate the principal stress.

    Parameters:
        sigma_x (float): Normal stress in the x direction.
        sigma_y (float): Normal stress in the y direction.
        tau_xy (float): Shear stress in the x-y plane.
        stress_type (str): 'max' for major principal stress, 'min' for minor principal stress.

    Returns:
        tuple: Major and minor principal stresses.
    """
    sigma_1 = (sigma_x + sigma_y)/2 + math.sqrt(((sigma_x - sigma_y)/2)**2 + tau_xy**2)
    sigma_2 = (sigma_x + sigma_y)/2 - math.sqrt(((sigma_x - sigma_y)/2)**2 + tau_xy**2)
    return sigma_1, sigma_2

def principal_stress_angle(sigma_x, sigma_y, tau_xy):
    """
    Calculate the principal stress angle.

    Parameters:
        sigma_x (float): Normal stress in the x direction.
        sigma_y (float): Normal stress in the y direction.
        tau_xy (float): Shear stress in the x-y plane.

    Returns:
        float: Principal stress angle in degrees.
    """
    numerator = 2 * tau_xy
    denominator = sigma_x - sigma_y
    theta_p_rad = math.atan2(numerator, denominator) / 2  # atan2 takes care of the angle quadrant
    theta_p_deg = math.degrees(theta_p_rad)  # converting radians to degrees
    
    return theta_p_deg, theta_p_deg + 90

def maximum_in_plane_shear_stress(sigma_x, sigma_y, tau_xy):
    """
    Calculate the maximum in-plane shear stress.

    Parameters:
        sigma_x (float): Normal stress in the x direction.
        sigma_y (float): Normal stress in the y direction.
        tau_xy (float): Shear stress in the x-y plane.

    Returns:
        float: Maximum in-plane shear stress.
    """
    tau_max = math.sqrt(((sigma_x - sigma_y)/2)**2 + tau_xy**2)
    return tau_max

def maximum_in_plane_shear_stress_angle(sigma_x, sigma_y, tau_xy):
    """
    Calculate the maximum in-plane shear stress angle.

    Parameters:
        sigma_x (float): Normal stress in the x direction.
        sigma_y (float): Normal stress in the y direction.
        tau_xy (float): Shear stress in the x-y plane.

    Returns:
        float: Maximum in-plane shear stress angle in degrees.
    """
    numerator = sigma_y - sigma_x
    denominator = 2 * tau_xy
    theta_max_rad = math.atan2(numerator, denominator) / 2  # atan2 takes care of the angle quadrant
    theta_max_deg = math.degrees(theta_max_rad)  # converting radians to degrees
    
    return theta_max_deg, theta_max_deg + 90

def mohrs_circle(sigma_x, sigma_y, tau_xy):
    """
    Calculate the center and radius of the Mohr's circle.

    Parameters:
        sigma_x (float): Normal stress in the x direction.
        sigma_y (float): Normal stress in the y direction.
        tau_xy (float): Shear stress in the x-y plane.

    Returns:
        tuple: (center, radius) of the Mohr's circle.
    """
    center = (sigma_x + sigma_y)/2
    radius = math.sqrt(((sigma_x - sigma_y)/2)**2 + tau_xy**2)
    return (center, radius)

def mohrs_circle_plot(sigma_x, sigma_y, tau_xy):
    """
    Plot the Mohr's circle.

    Parameters:
        sigma_x (float): Normal stress in the x direction.
        sigma_y (float): Normal stress in the y direction.
        tau_xy (float): Shear stress in the x-y plane.

    Returns:
        matplotlib.pyplot.plot: Mohr's circle plot.
    """
    
    center, radius = mohrs_circle(sigma_x, sigma_y, tau_xy)
    circle = plt.Circle((center, 0), radius, fill=False)
    fig, ax = plt.subplots()
    ax.set_aspect('equal')
    ax.add_artist(circle)
    
    # Set the x and y limits
    ax.set_xlim([center - radius - 0.1 * radius, center + radius + 0.1 * radius])  # adding a little extra space
    ax.set_ylim([-radius - 0.1 * radius, radius + 0.1 * radius])  # adding a little extra space
    
    plt.xlabel("Normal Stress")
    plt.ylabel("Shear Stress")
    plt.title("Mohr's Circle")
    plt.grid()
    plt.show()
    return fig

def mohrs_circle_stress(sigma_x, sigma_y, tau_xy, theta):
    """
    Calculate the normal and shear stresses on an inclined plane.

    Parameters:
        sigma_x (float): Normal stress in the x direction.
        sigma_y (float): Normal stress in the y direction.
        tau_xy (float): Shear stress in the x-y plane.
        theta (float): Angle of the inclined plane in degrees.

    Returns:
        tuple: (sigma_n, tau_n) on the inclined plane.
    """
    theta_rad = math.radians(theta)  # converting degrees to radians
    center, radius = mohrs_circle(sigma_x, sigma_y, tau_xy)
    sigma_n = center + radius*math.cos(2*theta_rad)
    tau_n = radius*math.sin(2*theta_rad)
    return (sigma_n, tau_n)

def mohrs_circle_stress_plot(sigma_x, sigma_y, tau_xy, theta):
    """
    Plot the normal and shear stresses on an inclined plane on the Mohr's circle.

    Parameters:
        sigma_x (float): Normal stress in the x direction.
        sigma_y (float): Normal stress in the y direction.
        tau_xy (float): Shear stress in the x-y plane.
        theta (float): Angle of the inclined plane in degrees.

    Returns:
        matplotlib.pyplot.plot: Normal and shear stresses on the inclined plane on the Mohr's circle.
    """
    sigma_n, tau_n = mohrs_circle_stress(sigma_x, sigma_y, tau_xy, theta)
    center, radius = mohrs_circle(sigma_x, sigma_y, tau_xy)
    circle = plt.Circle((center, 0), radius, fill=False)
    fig, ax = plt.subplots()
    ax.set_aspect('equal')
    ax.add_artist(circle)

    # Set the x and y limits
    ax.set_xlim([center - radius - 0.1 * radius, center + radius + 0.1 * radius])  # adding a little extra space
    ax.set_ylim([-radius - 0.1 * radius, radius + 0.1 * radius])  # adding a little extra space

    plt.plot(sigma_n, tau_n, 'ro')
    plt.xlabel("Normal Stress")
    plt.ylabel("Shear Stress")
    plt.title("Mohr's Circle")
    plt.grid()
    plt.show()
    return fig

def mohrs_circle_plane_angle(sigma_x, sigma_y, tau_xy, theta):
    """
    Calculate the angle of the plane that has the normal and shear stresses on the Mohr's circle.

    Parameters:
        sigma_x (float): Normal stress in the x direction.
        sigma_y (float): Normal stress in the y direction.
        tau_xy (float): Shear stress in the x-y plane.
        theta (float): Angle of the inclined plane in degrees.

    Returns:
        float: Angle of the plane that has the normal and shear stresses on the Mohr's circle in degrees.
    """
    sigma_n, tau_n = mohrs_circle_stress(sigma_x, sigma_y, tau_xy, theta)
    theta_rad = math.atan2(tau_n, sigma_n) / 2  # atan2 takes care of the angle quadrant
    theta_deg = math.degrees(theta_rad)  # converting radians to degrees
    return theta_deg

def mohrs_circle_plane_angle_plot(sigma_x, sigma_y, tau_xy, theta):
    """
    Plot the angle of the plane that has the normal and shear stresses on the Mohr's circle.

    Parameters:
        sigma_x (float): Normal stress in the x direction.
        sigma_y (float): Normal stress in the y direction.
        tau_xy (float): Shear stress in the x-y plane.
        theta (float): Angle of the inclined plane in degrees.

    Returns:
        matplotlib.pyplot.plot: Angle of the plane that has the normal and shear stresses on the Mohr's circle.
    """
    theta_deg = mohrs_circle_plane_angle(sigma_x, sigma_y, tau_xy, theta)
    sigma_n, tau_n = mohrs_circle_stress(sigma_x, sigma_y, tau_xy, theta)
    center, radius = mohrs_circle(sigma_x, sigma_y, tau_xy)
    circle = plt.Circle((center, 0), radius, fill=False)
    fig, ax = plt.subplots()
    ax.set_aspect('equal')
    ax.add_artist(circle)

    # Set the x and y limits
    ax.set_xlim([center - radius - 0.1 * radius, center + radius + 0.1 * radius])  # adding a little extra space
    ax.set_ylim([-radius - 0.1 * radius, radius + 0.1 * radius])  # adding a little extra space

    plt.plot(sigma_n, tau_n, 'ro')
    plt.plot([center, sigma_n], [0, tau_n], 'k-')
    plt.plot([center, sigma_n], [0, 0], 'k-')
    plt.plot([sigma_n, sigma_n], [0, tau_n], 'k-')
    plt.xlabel("Normal Stress")
    plt.ylabel("Shear Stress")
    plt.title("Mohr's Circle")
    plt.grid()
    plt.show()
    return fig

